src/WebJobs.Script.WebHost/Diagnostics/DiagnosticEventTableStorageRepository.cs

```diff
@@ -101,43 +109,96 @@ internal CloudTable GetDiagnosticEventsTable(DateTime? now = null)
+            return $"{TableNamePrefix}{date:yyyyMM}";
         }
 
         protected internal virtual async void OnFlushLogs(object state)
         {
             await FlushLogs();
         }
 
+        private async Task PurgePreviousEventVersions()
+        {
+            _logger.LogTrace("Purging diagnostic events with versions older than '{currentEventVersion}'.", DiagnosticEvent.CurrentEventVersion);
+
+            await Utility.InvokeWithRetriesAsync(async () =>
+            {
+                try
+                {
+                    var tables = (await TableStorageHelpers.ListTablesAsync(TableClient, TableNamePrefix)).ToList();
+
+                    foreach (var table in tables)
+                    {
+                        var tableRecords = await table.ExecuteQuerySegmentedAsync(new TableQuery<DiagnosticEvent>(), null);
+
+                        // Skip tables that have 0 records
+                        if (tableRecords.Results.Count == 0)
+                        {
+                            continue;
+                        }
+
+                        // Delete table if it doesn't have records with EventVersion
+                        var eventVersionDoesNotExists = tableRecords.Results.Any(record => string.IsNullOrEmpty(record.EventVersion) == true);
+                        if (eventVersionDoesNotExists)
+                        {
+                            _logger.LogTrace("Deleting table '{tableName}' as it contains records without an EventVersion.", table.Name);
+                            await table.DeleteIfExistsAsync();
+                            continue;
+                        }
+
+                        // If the table does have EventVersion, query if it is an outdated version
+                        var eventVersionOutdated = tableRecords.Results.Any(record => string.Compare(DiagnosticEvent.CurrentEventVersion, record.EventVersion, StringComparison.Ordinal) > 0);
+                        if (eventVersionOutdated)
+                        {
+                            _logger.LogTrace("Deleting table '{tableName}' as it contains records with an outdated EventVersion.", table.Name);
+                            await table.DeleteIfExistsAsync();
+                        }
+                    }
+
+                    _purged = true;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Error occurred when attempting to purge previous diagnostic event versions.");
+                }
+            }, maxRetries: 5, retryInterval: TimeSpan.FromSeconds(5));
+        }
+
         internal virtual async Task FlushLogs(CloudTable table = null)
         {
             if (_environment.IsPlaceholderModeEnabled())
             {
                 return;
             }
 
+            if (_primaryHostStateProvider.IsPrimary && !_purged)
+            {
+                _ = PurgePreviousEventVersions();
```

Comment:
If we kick this off... we probably want to at least sleep here for 30 seconds or so? Otherwise, I can imagine a race where the code below thinks the table exists and starts writing to it... only for it to be deleted out from under it. Waiting 30 seconds or so if we *know* we're deleting a table would calm that down.

Reply:
We don't delete empty tables so this should be fine? Can add a 30 second delay if still worried

Reply:
Added 30s delay if a table is deleted, also changed this to no longer be a background task and is now awaited

Reply:
Yeah, I was thinking of this mixed-version case. Over time this wouldn't really be an issue (until we change versions again). 

And I think you're right that awaiting is a good approach. There's no rush to flush these (we've already waited 10 minutes) so being more deterministic works for us in this case.